File: scripts/python/sacred_cli/window.py
"""Window wrapper. Mirrors scripts/cli/lib/window.js."""

from .ansi import RESET, COLORS, bg_default, bg_hex, vis_len

MARGIN = 2
SHADOW_W = 1
SHADOW_H = 1

MIN_INNER_W = 20
MIN_TERM_W = 2 * MARGIN + SHADOW_W + MIN_INNER_W

bg_term = bg_default
bg_win = bg_hex(COLORS["windowBg"])
bg_shd = bg_hex(COLORS["shadow"])
margin_right = " " * max(0, MARGIN - SHADOW_W)


def get_inner_width(term_w: int) -> int:
    return max(10, term_w - 2 * MARGIN - SHADOW_W)


def wrap_line(line: str, inner_w: int) -> str:
    pad = inner_w - vis_len(line)
    fill = " " * pad if pad > 0 else ""
    return (
        f"{bg_term}{' ' * MARGIN}{bg_win}{line}{RESET}"
        f"{bg_win}{fill}{RESET}{bg_shd} {RESET}{bg_term}{margin_right}{RESET}"
    )


def shadow_bottom_row(inner_w: int) -> str:
    return (
        f"{bg_term}{' ' * (MARGIN + SHADOW_W)}{bg_shd}{' ' * inner_w}{RESET}"
        f"{bg_term}{margin_right}{RESET}"
    )


def wrap_line_top(line: str, inner_w: int) -> str:
    pad = inner_w - vis_len(line)
    fill = " " * pad if pad > 0 else ""
    return (
        f"{bg_term}{' ' * MARGIN}{bg_win}{line}{RESET}"
        f"{bg_win}{fill}{RESET}{bg_term}{' ' * MARGIN}{RESET}"
    )


def wrap_lines(lines, term_w: int):
    inner_w = get_inner_width(term_w)
    out = []
    if lines:
        out.append(wrap_line_top(lines[0], inner_w))
    for line in lines[1:]:
        out.append(wrap_line(line, inner_w))
    out.append(shadow_bottom_row(inner_w))
    return out

File: scripts/cli/lib/window.ts
import { bgHex, bgDefault, RESET, COLORS, visLen } from './ansi';

const MARGIN: number = 2;
const SHADOW_W: number = 1;
const SHADOW_H: number = 1;

const MIN_INNER_W: number = 20;
const MIN_TERM_W: number = 2 * MARGIN + SHADOW_W + MIN_INNER_W;

const bgTerm: string = bgDefault;
const bgWin: string = bgHex(COLORS.windowBg);
const bgShd: string = bgHex(COLORS.shadow);
const marginRight: string = ' '.repeat(Math.max(0, MARGIN - SHADOW_W));

function getInnerWidth(termW: number): number {
  return Math.max(10, termW - 2 * MARGIN - SHADOW_W);
}

function wrapLine(line: string, innerW: number): string {
  const pad = innerW - visLen(line);
  const fill = pad > 0 ? ' '.repeat(pad) : '';
  return `${bgTerm}${' '.repeat(MARGIN)}${bgWin}${line}${RESET}${bgWin}${fill}${RESET}${bgShd} ${RESET}${bgTerm}${marginRight}${RESET}`;
}

function shadowBottomRow(innerW: number): string {
  return `${bgTerm}${' '.repeat(MARGIN + SHADOW_W)}${bgShd}${' '.repeat(innerW)}${RESET}${bgTerm}${marginRight}${RESET}`;
}

function wrapLineTop(line: string, innerW: number): string {
  const pad = innerW - visLen(line);
  const fill = pad > 0 ? ' '.repeat(pad) : '';
  return bgTerm + ' '.repeat(MARGIN) + bgWin + line + RESET + bgWin + fill + RESET + bgTerm + ' '.repeat(MARGIN) + RESET;
}

function wrapLines(lines: string[], termW: number): string[] {
  const innerW = getInnerWidth(termW);
  const out: string[] = [];
  if (lines.length > 0) out.push(wrapLineTop(lines[0], innerW));
  for (let i = 1; i < lines.length; i++) out.push(wrapLine(lines[i], innerW));
  out.push(shadowBottomRow(innerW));
  return out;
}

export {
  MARGIN, SHADOW_W, SHADOW_H, MIN_INNER_W, MIN_TERM_W,
  bgTerm, bgWin, bgShd, marginRight,
  getInnerWidth, wrapLine, wrapLineTop, shadowBottomRow, wrapLines,
};
